File: wick/step_6a_linked_cluster.py
"""
Step 6a: Linked-Cluster Theorem Filter

Remove unlinked (disconnected) diagram contributions.

In MBPT, the linked-cluster theorem states that unlinked diagram contributions
to the energy cancel exactly. A diagram is LINKED if all interaction vertices
(tensors) are connected via shared indices.

Algorithm:
  - Build a connectivity graph where nodes = tensors, edges = shared indices
  - Check if all tensors form one connected component
  - Drop terms with multiple disconnected pieces

Example disconnected terms (REMOVED):
  - <i,j||i,j> * <k,l||k,l>          (two separate bubbles)
  - <i,j||a,b> * <k,l||k,l>          ((i,j,a,b) separate from (k,l))
  - <i,j||i,j> * <a,b||a,b> * <c,d||c,d>  (three separate pieces)

Example connected terms (KEPT):
  - <i,j||a,b> * <a,b||i,j>          (i,j,a,b all connected)
  - <i,j||a,b> * <a,k||b,l> * <k,l||i,j>  (all connected through indices)
"""

from __future__ import annotations
from typing import List
from .term import Term


def _tensors_share_index(t1, t2) -> bool:
    """Check if two tensors share at least one index."""
    return not set(t1.indices).isdisjoint(t2.indices)


def _is_connected(term: Term) -> bool:
    """Check if all tensors in a term are connected via shared indices."""
    tensors = term.tensors
    if len(tensors) <= 1:
        return True

    # BFS connectivity check: can we reach all tensors from tensor 0?
    visited = {0}
    stack = [0]

    while stack:
        i = stack.pop()
        for j in range(len(tensors)):
            if j in visited:
                continue
            if _tensors_share_index(tensors[i], tensors[j]):
                visited.add(j)
                stack.append(j)

    return len(visited) == len(tensors)


def filter_unlinked_diagrams(terms: List[Term]) -> List[Term]:
    """Keep only linked (connected) diagrams."""
    result = []
    for term in terms:
        if _is_connected(term):
            result.append(term)
    return result

File: src/step_6a_linked_cluster.rs
//! Step 6a: Linked-Cluster Theorem Filter
//!
//! Remove unlinked (disconnected) diagram contributions.
//! A diagram is LINKED if all tensors are connected via shared indices.

use std::collections::HashSet;

use crate::term::Term;

fn tensors_share_index(t1: &crate::term::Tensor, t2: &crate::term::Tensor) -> bool {
    let s1: HashSet<&String> = t1.indices.iter().collect();
    t2.indices.iter().any(|idx| s1.contains(idx))
}

fn is_connected(term: &Term) -> bool {
    let tensors = &term.tensors;
    if tensors.len() <= 1 {
        return true;
    }

    let mut visited = HashSet::new();
    visited.insert(0usize);
    let mut stack = vec![0usize];

    while let Some(i) = stack.pop() {
        for j in 0..tensors.len() {
            if visited.contains(&j) {
                continue;
            }
            if tensors_share_index(&tensors[i], &tensors[j]) {
                visited.insert(j);
                stack.push(j);
            }
        }
    }

    visited.len() == tensors.len()
}

pub fn filter_unlinked_diagrams(terms: &[Term]) -> Vec<Term> {
    terms.iter().filter(|t| is_connected(t)).cloned().collect()
}
